File: TP_MPI/occurrences.py
#!/bin/python3
# -*- coding: utf-8 -*-

# Programme qui compte les occurrences des lettres dans un texte
# Execution : invocation directe dans un terminal ou via python3

# Bibliothèques utilisées
import sys
import math
import random as rnd
import string

# Import de MPI
from mpi4py import MPI

#
# Constantes utilisées dans le programme
#
NB_LETTRES = 26

#
# Génération pseudo-aléatoire d'un texte de n caractères
#
def genereTexte(n, graine):
    texte = ""

    # Init PRNG
    rnd.seed(graine)
    # Ensemble des caractères affichables (lettres, ponctuations,...)
    taille = len(string.printable)
    
    # Tirage aléatoire de chaque caractère dans l'ensemble affichable
    for i in range(n):
        texte += string.printable[rnd.randint(0,taille-1)]

    return texte

#
# Version séquentielle du comptage des occurrences
#
def compte_occs_seq(texte, n):
    # Tableau des occurrences
    occus = [0 for i in range(NB_LETTRES)]
    # Parcours du texte
    for car in texte:
        if (car in string.ascii_lowercase):
            occus[ord(car) - ord('a')] += 1
        elif (car in string.ascii_uppercase):
            occus[ord(car) - ord('A')] += 1

    return occus

#
# Version parallèle du comptage des lettres avec un schéma maître-travailleurs
# !! CETTE FONCTION DOIT ÊTRE MODIFIÉE !!
#
def compte_occs_par(texte, n, comm, idP, nbP, block):
    # Tableau des occurrences
    occus = [0 for i in range(NB_LETTRES)]
    # Identifiant du maître
    maitre = 0

    # Processus maître
    if (idP == maitre):
        
        # Variables locales
        reste      = n # Nombre de caractères restant à traiter
        position   = 0 # Position du 1er caractère du sous-texte
        nbAEnvoyer = 0 # Nombre de caractères à envoyer
        nbTaches   = 0 # Nombre de tâches actives

        # 1ère distribution des blocs de texte (1 pour chaque travailleur)
        for i in range(1,nbP):
            
            # Nombre de caractères à envoyer au travailleur i
            # ...

            # Envoi du bloc au travailleur i avec l'étiquette 0
            # ...

            # Mise à jour des informations globales
            # ...

            # Affichage des informations (DOIT ÊTRE COMMENTÉ PENDANT LES MESURES DE PERFS)
            print("Envoi d'un bloc de taille %d au processus %d" % (nbAEnvoyer, i))

        # Boucle de réception des résultats et envoi d'autres travaux tant qu'il en reste
        while (nbTaches > 0):

            # Variable contenant les infos d'état de la réception
            info = MPI.Status()

            # Attente d'un résultat de n'importe quel travailleur
            # ...
            
            # Mise à jour du nombre de tâches en cours
            # ...
            
            # Mise à jour du résultat global
            # ...

            # Envoi du bloc de texte suivant s'il en reste
            if (reste > 0):
               pass # LIGNE À SUPPRIMER LORSQUE LES LIGNES CI-DESSOUS SONT COMPLÉTÉES

               # Nombre de caractères à envoyer au travailleur i
               # ...

               # Envoi du block au MÊME travailleur avec l'étiquette 0
               # ...

               # Affichage des informations (DOIT ÊTRE COMMENTÉ PENDANT LES MESURES DE PERFS)
               # print("Envoi d'un bloc de taille %d au processus %d" % (nbAEnvoyer, ...))

               # Mise à jour des infos globales
               # ...

        # Tout le texte a été traité
        # Envoi du signal de terminaison aux travailleurs
        # ...
        
          #
    else: # Travailleurs
          #

        # Variables locales
        fini    = False
        nbRecus = 0

        # Boucle de travail
        while (not fini):
            
            # Attente d'un travail du maître avec l'étiquette 0
            # ...

            # Nombre de caractères reçus
            # ...

            # Un message vide correspond au signal de termination 
            if (nbRecus == 0):
                # Indique la fin du travail
                fini = True
            else:
                pass # LIGNE À SUPPRIMER LORSQUE LES LIGNES CI-DESSOUS SONT COMPLÉTÉES

                # Comptage des occurrences par le travailleur courant
                # ...

                # Envoi du résultat au maître avec l'étiquette 1
                # ...

    return occus

#
# Comparaison des contenus des tableaux de résultats : retourne True si identiques
#
def compare_tableaux(arrayA, arrayB):
    if (len(arrayA) != len(arrayB)):
        return False
    for i in range(len(arrayA)):
        if (arrayA[i] != arrayB[i]):
            return False
    return True

#
# Fonction principale
#
def main():

    # Valeurs par défaut des paramètres du programme
    n      = 1000000 # Taille du texte
    bloc   = 1000    # Taille d'un bloc
    graine = 1       # Graine du PRNG

    # Paramètres de la ligne de commande
    argc = len(sys.argv)
    for i in range(argc):
        if sys.argv[i] == "-b":
            i = i + 1
            bloc = int(sys.argv[i])
        elif sys.argv[i] == "-t":
            i = i + 1
            n = int(sys.argv[i])
        elif sys.argv[i] == "-g":
            i = i + 1
            graine = int(sys.argv[i])
    
    # Variables internes du programme
    texte      = None # Texte
    tdeb       = 0.0  # Temps de début
    tfin       = 0.0  # Temps de fin
    tSeq       = 0.0  # Temps séquentiel
    tPar       = 0.0  # Temps parallèle
    accel      = 0.0  # Accélération
    efficacite = 0.0  # Efficacité

    # Init MPI
    comm = MPI.COMM_WORLD  # Groupe de communication
    idP  = comm.Get_rank() # Identifiant du processus courant
    nbP  = comm.Get_size() # Nombre de processus dans le groupe

    # Affichage des paramètres du programme par le processus 0 (maître)
    if (idP == 0):
        print()
        print("Taille texte : %d" % (n))
        print("Taille blocs : %d" % (bloc))
        print("Graine       : %u" % (graine))
        print("Nb processus : %d" % (nbP))
        print()

    # Génération du texte par le processus 0 (maître)
    if(idP == 0):
        texte = genereTexte(n, graine)
    
    #
    # Version SÉQUENTIELLE du comptage des occurrences par le processus 0 (maître)
    #
    if (idP == 0):
        # Temps début
        tdeb = MPI.Wtime()
        # Comptage séquentiel
        occsSeq = compte_occs_seq(texte, n)
        # Temps fin
        tfin = MPI.Wtime()
        # Temps total
        tSeq = tfin - tdeb

    #
    # Version PARALLÈLE du comptage des occurrences avec un schéma Maître-Travailleurs (maître = 0)
    #
    # Temps début
    tdeb = MPI.Wtime()
    # Comptage séquentiel
    occsPar = compte_occs_par(texte, n, comm, idP, nbP, bloc)
    # Temps fin
    tfin = MPI.Wtime()
    # Temps total
    tPar = tfin - tdeb

    # Validation des résultats et affichage des informations par le maître
    if (idP == 0):
        print()

        # Validation (comparaison)       
        if (compare_tableaux(occsSeq, occsPar)):
            print("Résultats IDENTIQUES entre les versions séquentielle et parallèle !")
        else:
            print("Résultats DIFFÉRENTS entre les versions séquentielle et parallèle !")
        print()

        # Affichage des occurrences des lettres dans le texte
        print("       Seq      Par     Fréquence");
        for i in range(NB_LETTRES):
            print("%c = %8d %8d    %.2e" % (ord('a')+i, occsSeq[i], occsPar[i], occsSeq[i]/n))
        print()

        # Évaluation des performances
        accel      = tSeq / tPar
        efficacite = accel / (nbP - 1)
        print("Temps séquentiel : %f s\nTemps parallèle  : %f s\nAccélération     : %f\nEfficacité       : %f" % (tSeq, tPar, accel, efficacite))

#
# Programme principal
#
if  __name__  == "__main__":
    main()

File: TP_MPI/occurrences.c
/* 
   Programme de comptabilisation des occurrences des lettres dans un texte
   Fichier à compiler avec mpicc
   
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <time.h>
#include <sys/time.h>
#include <math.h>
#include <mpi.h>

//
// Déclaration des constantes utilisées dans le programme
//
const int NB_LETTRES = 26;

//
// Déclarations des fonctions utilisées dans le programme
//
char *genere_texte(int taille, uint graine);                                     // Génération d'un texte aléatoire
void compte_occurrences_seq(char *texte, int taille, int *occus);                // Comptage des occurrences en séquentiel
void compte_occurrences_mpi(int numP, int nbP, int tBloc, char *texte, int taille, int *occus); // Comptage des occurrences en MPI
char tabsIdem(int OccurrencesA[], int OccurrencesB[]);                           // Comparaison de deux tableaux de nombres d'occurrences

//
// Programme principal
//
int main(int argc, char **argv)
{
  //
  // Variables 
  //
  char  *texte  = NULL;         // Texte à traiter
  int    taille = 10000;        // Taille du texte à traiter
  int    tBloc  = 1000;         // Taille des blocs de texte envoyés aux travailleurs
  double deb;                   // Chrono pour les mesures de temps
  double tempsSeq = 0;          // Temps de calcul séquentiel
  double tempsMPI = 0;          // Temps de calcul MPI
  double accelMPI, effMPI;      // Accélaration et efficacité MPI
  int    opt;                   // Option de ligne de commande
  uint   graine = time(NULL);   // Graine du générateur aléatoire initialisée avec l'horloge de la machine
  int    numP, nbP;             // Numéro du processus MPI et nombre de processus MPI
  int occurrencesS[NB_LETTRES]; // Occurrences des lettres obtenues en séquentiel
  int occurrencesM[NB_LETTRES]; // Occurrences des lettres obtenues en MPI

  //
  // Lecture des paramètres
  //
  while ((opt = getopt(argc, argv, "b:g:t:")) != -1) {
    switch (opt) {
    case 'b':                 // Taille des blocs MPI
      tBloc = atoi(optarg);
      break;
    case 'g':                 // Graine
      graine = atoi(optarg);
      break;
    case 't':                 // Taille du texte
      taille = atoi(optarg);
      break;
    }
  }

  // Init MPI
  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nbP);  // Récupération du nombre de processus MPI
  MPI_Comm_rank(MPI_COMM_WORLD, &numP); // Récupération du numéro du processus MPI courant

  //
  // Affichage des paramètres du programme
  //
  if(numP == 0){
    printf("Taille du texte      : %d\n", taille);
    printf("Graine               : %u\n", graine);
    printf("Nb Procs MPI         : %d\n", nbP);
    printf("Taille des blocs MPI : %d\n", tBloc);
  }

  //
  // Construction et initialisation aléat du texte
  // 
  srand48(graine); // Initialisation du générateur aléatoire avec la graine donnée
  
  // Génération du texte uniquement sur le processus MPI 0
  if(numP == 0){

    // Initialisation des tableaux d'occurrences 
    memset(occurrencesS, 0, NB_LETTRES * sizeof(int));
    memset(occurrencesM, 0, NB_LETTRES * sizeof(int));

    texte = genere_texte(taille, graine);
  }

  //
  // Comptage des occurrences en séquentiel sur proc MPI 0
  //
  if(numP == 0){
    deb = MPI_Wtime();
    compte_occurrences_seq(texte, taille, occurrencesS);
    tempsSeq = MPI_Wtime() - deb;
    printf("Temps Séq : %f s\n", tempsSeq);
  }

  //
  // Comptage des occurrences en MPI via un schéma maître-travailleurs piloté par le proc 0
  //
  deb = MPI_Wtime();
  compte_occurrences_mpi(numP, nbP, tBloc, texte, taille, occurrencesM);
  tempsMPI = MPI_Wtime() - deb;
  
  // Validation des résultats et affichages par le proc 0
  if(numP == 0){

    // Validation des calculs parallèles (comparaison des résultats)
    if(tabsIdem(occurrencesS, occurrencesM) == 'O'){
      printf("Les résultats séquentiel et MPI sont identiques !\n");
    }else{
      printf("Les résultats séquentiel et MPI sont DIFFÉRENTS :-/\n");
    }

    // Mesure de la qualité du parallélisme MPI
    accelMPI = tempsSeq / tempsMPI;
    effMPI = accelMPI / (nbP - 1);
    printf("MPI :\n\tTemps        : %f s\n\tAccélération : %f\n\tEfficacité   : %f\n", tempsMPI, accelMPI, effMPI);

    // Affichage des occurrences des lettres dans le texte
    printf("       Séq      MPI     Fréquences\n");
    int i;
    for(i=0; i<NB_LETTRES; ++i){
      printf("%c = %8d %8d\t%.2e\n", 'a'+i, occurrencesS[i], occurrencesM[i], (float)occurrencesS[i]/taille);
    }
  }

  // Arrêt du système MPI
  MPI_Finalize();

  //
  // Destruction des tableaux dynamiques
  //  
  free(texte);
  
  // Fin
  return 0;
}

//
// Génération du texte
//
char *genere_texte(int taille, uint graine)
{
  char *texte = malloc(taille);
  int i;

  printf("Génération du texte...");
  fflush(stdout);
  #pragma omp parallel for
  for(i=0; i<taille; ++i){
    texte[i] = (rand_r(&graine) % 95) + ' '; // 95 caractères affichables à partir de l'espace
  }
  printf("terminée.\n");
  return texte;
}

//
// Comptage des occurrences en séquentiel
//
void compte_occurrences_seq(char *texte, int taille, int *occus)
{
  int i;

  for(i=0; i<taille; ++i){
    if(texte[i] >= 'a' && texte[i] <= 'z'){
      occus[texte[i]-'a']++;
    }
    if(texte[i] >= 'A' && texte[i] <= 'Z'){
      occus[texte[i]-'A']++;
    }
  }
}

//
// Comptage des occurrences en MPI en mode client-serveur piloté par le processus 0
//
void compte_occurrences_mpi(int numP, int nbP, int tBloc, char *texte, int taille, int *occus)
{
  int i;               // Indice de boucle
  char *tampon = NULL; // Tampon mémoire de travail
  MPI_Status etat;     // État de réception

  // Allocation du tampon local
  tampon = (char *) malloc(tBloc);
  
  // Gestion du maître (proc 0)
  if(numP == 0){
  
    int nbRestants = taille; // Nombre de caractères restant à traiter dans le texte
    int src;                 // Source du message reçu
    int nbEnvoyes;           // Nombre de caractères envoyés
    int occs[NB_LETTRES];    // Nombre d'occurrences
    int nbAttendus = 0;      // Nombre de résultats attendus (tâches en cours)

    // 1ère distribution du maître (proc 0)
    for(i=1; i<nbP; ++i){

      // Calcul du nombre de caractères à envoyer
      // ...

      // Copie dans tampon et envoi au processus cible
      // ...

    }

    // Boucle de gestion du travail... tant qu'il en reste
    while(nbAttendus > 0){

      // Récupération d'un résultat (réception d'un message d'un autre processus)
      // ...
      
      // Mise à jour du résultat global
      // ...

      // Envoi du bloc suivant au processus libre s'il reste du travail
      // ...

    }

    // Signal de fin (message vide) envoyé à chaque travailleur
    // ...

  }else{ // Gestion des travailleurs

    int oloc[NB_LETTRES]; // Tableau local des occurrences des lettres
    int nbRecus = 0;      // Nombre de caractères reçus
    char fini = 'N';      // Indicateur de fin du travail

    while(fini == 'N'){

      // Attente d'un message du maître
      // ...

      // Récupération de la taille du message
      // ...
      
      // Distinction entre message de texte et message de fin (vide)
      if(nbRecus == 0){ // Taille nulle indique la fin du travail

        // Arrêt du travail
        fini = 'O';
        
      }else{ // Taille non nulle indique un morceau de texte à traiter

        // Réception du texte
        // ...

        // Initialisation des nombres d'occurrences locaux
        // ...
        
        // Traitement du texte
        // ...

        // Envoi du résultat au maître
        // ...

      }
    }
  }

  // Libération du tampon local
  free(tampon);
}

//
// Comparaison de deux tableaux d'occurrences de lettres
//
char tabsIdem(int OccurrencesA[], int OccurrencesB[])
{
  int i;

  for(i=0; i<NB_LETTRES && OccurrencesA[i] == OccurrencesB[i]; ++i);
  
  if(i < NB_LETTRES){
    return 'N';
  }else{
    return 'O';
  }
}
